modules/fledge.js

```diff
@@ -0,0 +1,192 @@
+let _componentAuctions = [];
+
+config.getConfig('fledge', config => init(config.fledge));
+
+/**
+  * Module init.
+  */
+export function init(fledgeCfg) {
+  logInfo(MODULE, `init`, fledgeCfg);
+  if (fledgeCfg.enabled !== false) {
+    if (fledgeCfg.useGpt) {
+      getHook('addComponentAuction').before(addComponentAuctionToGptHook);
+    } else {
+      getHook('addComponentAuction').before(addComponentAuctionHook);
+      getHook('renderAd').before(renderAdHook);
+      getHook('secureRenderRequest').before(secureRenderRequestHook);
+    }
+  }
+}
+
+export function addComponentAuctionToGptHook(next, bidRequest, componentAuctionConfig) {
+  const seller = componentAuctionConfig.seller;
+  const adUnitCode = bidRequest.adUnitCode;
+  const gptSlot = getGptSlotForAdUnitCode(adUnitCode);
+  logWarn(MODULE, 'gptSlotInfo', gptSlot);
+  if (gptSlot && gptSlot.setConfig) {
+    delete componentAuctionConfig.bidId;
+    gptSlot.setConfig({
+      componentAuction: [{
+        configKey: seller,
+        auctionConfig: componentAuctionConfig
+      }]
+    });
+    logInfo(MODULE, `register component auction config in GPT for: ${adUnitCode} x ${seller}: ${gptSlot.getAdUnitPath()}`, componentAuctionConfig);
+  } else {
+    logWarn(MODULE, `unable to register component auction config in GPT for: ${adUnitCode} x ${seller}.`);
+  }
+
+  next(bidRequest, componentAuctionConfig);
+}
+
+export function addComponentAuctionHook(next, bidRequest, fledgeAuctionConfig) {
+  if (!bidRequest || !fledgeAuctionConfig) { return; }
+
+  const seller = fledgeAuctionConfig.seller;
+  const adUnitCode = bidRequest.adUnitCode;
+
+  // TODO : remove duplicate {adUnitCode, seller} is already registered
+  logInfo(MODULE, `adding auction config for ${adUnitCode} x ${seller}`);
+  _componentAuctions.push({adUnitCode, seller, fledgeAuctionConfig});
+
+  next(bidRequest, fledgeAuctionConfig);
+}
+
+export function getAuctionConfig(bid) {
+  const targetAdUnitCode = bid.adUnitCode;
+  const componentAuctionsForAdUnitCode = _componentAuctions.filter((componentAuction) => {
+    const {adUnitCode} = componentAuction;
+    return targetAdUnitCode == adUnitCode;
+  });
+  if (componentAuctionsForAdUnitCode.length == 0) {
+    logInfo(MODULE, `NOT running runAdAuction for adUnitCode=${targetAdUnitCode} since there are no componentAuctions`);
+    return null;
+  }
+
+  logInfo(MODULE, `runAdAuction for adUnitCode=${targetAdUnitCode}`, componentAuctionsForAdUnitCode);
+
+  const seller = config.getConfig('fledge.seller');
+  const decisionLogicUrl = config.getConfig('fledge.decisionLogicUrl');
+  if (seller && decisionLogicUrl) {
+    // multiple-ssp with top level support
+    return {
+      seller,
+      decisionLogicUrl,
+      componentAuctions: componentAuctionsForAdUnitCode.map((auction) => auction.fledgeAuctionConfig)
+    }
+  }
+
+  // fallback if there is no top-level seller defined
+  const randomIdx = Math.floor(Math.random() * componentAuctionsForAdUnitCode.length);
```

Comment:
The expectation for a correctly configured publisher is for the `decisionLogicUrl` to be defined and the "multiple ssp with top level support" path be taken. The decision logic script will likely be provided by prebid as a common resource in some way.

Due to the special requirements for serving it*), it cannot be bundled in, so for the moment we made it configurable.

*) the response header `x-allow-fledge: true` must be passed when serving the decision logic scripts. It also has to come from the same domain as the seller.

Reply:
That's reasonable, but if no decision logic is available, to me it'd seem better to discard fledge rather than awarding the auction to a random fledge bidder.